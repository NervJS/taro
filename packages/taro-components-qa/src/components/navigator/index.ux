<template>
  <div
    class="{{classname}}"
    style="{{customstyle}}"
    id="{{id}}"
    onclick="handleClick"
  >
    <slot></slot>
  </div>
</template>

<script>
  import router from '@system.router'

  export default {
    props: {
      classname: {
        type: String,
        required: false,
        default: 'taro_navigator'
      },
      customstyle: {
        type: Object,
        required: false,
        default: {}
      },
      id: {
        type: String,
        required: false,
        default: ''
      },
      url: {
        type: String,
        required: true,
        default: ''
      },
      opentype: {
        type: String,
        required: false,
        default: 'navigate'
      }
    },

    data: () => ({
    }),
    onInit () {
    },
    handleClick (e) {
      let params = this.getUrlParams(this.url)
      switch (this.opentype) {
        case 'redirect':
          router.replace({uri: this.url, params})
          break
        case 'reLaunch':
          router.clear()
          router.replace({uri: this.url, params})
          break
        case 'navigateBack':
          router.back()
          break
        default:
          router.push({uri: this.url, params})

      }
      this.$emit('click', e)
    },
    getUrlParams (url = '') {
      let params = {}
      url && url.replace(/(\w+)=(\w+)/ig, function (a, b, c) {
        params[b] = unescape(c)
      })
      return params
    }
  }
</script>

<style>
  .taro_navigator {
    height: 30px;
  }
</style>
