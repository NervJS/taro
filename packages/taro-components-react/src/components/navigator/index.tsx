import './loader'

export { Navigator as default } from '@tarojs/components/lib/react'
