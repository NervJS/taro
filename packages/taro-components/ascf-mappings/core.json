{
  "Button": {
    "size": {
      "ascfProp": "dimension",
      "valueMap": {
        "mini": "XS",
        "small": "S",
        "default": "M",
        "large": "L"
      },
      "required": false,
      "default": "M",
      "docLink": "https://ascf.docs/components/button#dimension"
    },
    "loading": {
      "ascfProp": "isLoading",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/button#isloading"
    }
  },
  "Input": {
    "password": {
      "ascfProp": "secretInput",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/input#secretinput"
    }
  }
}
