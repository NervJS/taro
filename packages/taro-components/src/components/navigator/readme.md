# taro-navigator-core



<!-- Auto Generated Below -->


## Overview

TODO: 参数还需要进一步细化对齐
Navigator组件
https://developers.weixin.qq.com/miniprogram/dev/component/navigator.html

## Properties

| Property     | Attribute     | Description | Type      | Default      |
| ------------ | ------------- | ----------- | --------- | ------------ |
| `delta`      | `delta`       |             | `number`  | `0`          |
| `hoverClass` | `hover-class` |             | `string`  | `undefined`  |
| `isHover`    | `is-hover`    |             | `boolean` | `false`      |
| `openType`   | `open-type`   |             | `string`  | `'navigate'` |
| `url`        | `url`         |             | `string`  | `undefined`  |


## Events

| Event      | Description | Type               |
| ---------- | ----------- | ------------------ |
| `complete` |             | `CustomEvent<any>` |
| `cuccess`  |             | `CustomEvent<any>` |
| `fail`     |             | `CustomEvent<any>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
