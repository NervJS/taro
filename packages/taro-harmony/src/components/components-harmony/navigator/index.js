import router from '@system.router'
import { createOption, queryToJson } from '../utils'

export default createOption({
  props: {
    id: {
      default: ''
    },
    openType: {
      default: 'navigate'
    },
    url: {
      default: ''
    },
    hoverStartTime: {
      default: 50
    },
    hoverStayTime: {
      default: 600
    },
    hoverClass: {
      default: 'navigator-hover'
    }
  },

  data () {
    return {
      targetObj: {
        navigate: 'push',
        redirect: 'replace',
        switchTab: 'push',
        reLaunch: 'replace',
        navigateBack: 'back'
      },
      hover: false,
      touch: false
    }
  },

  computed: {
    clsHover () {
      return this.hover ? this.hoverClass : ''
    }
  },

  onClick () {
    const { openType = 'navigate', url = '' } = this
    this.getRouterFunc(`${openType}`, url)
  },

  getRouterFunc (method, url) {
    const methodName = this.targetObj[method]
    const [uri, queryString = ''] = url.split('?')
    const params = queryToJson(queryString)

    const uriObj = url ? { uri: uri.replace(/^\//, '') } : {}

    const paramsObj = queryString ? { params } : {}

    if (method === 'reLaunch') {
      return router.clear()
    } else if (method === 'switchTab') {
      const app = getApp()
      const pages = app.pageStack

      for (let i = 0; i < pages.length; i++) {
        const item = pages[i]
        if (item === uriObj.uri) {
          return router.back({
            uri: item
          })
        }
      }
    }

    router[methodName]({
      ...uriObj,
      ...paramsObj
    })
  },

  onTouchStart () {
    this.touch = true
    if (this.hoverClass) {
      setTimeout(() => {
        if (this.touch) {
          this.hover = true
        }
      }, this.hoverStartTime)
    }
  },

  onTouchEnd () {
    this.touch = false
    if (this.hoverClass) {
      setTimeout(() => {
        if (!this.touch) {
          this.hover = false
        }
      }, this.hoverStayTime)
    }
  }
})
