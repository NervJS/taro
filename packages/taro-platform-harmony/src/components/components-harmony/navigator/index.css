.navigate {
  overflow: hidden;
  transition: all linear 0.2s;
}

.navigator-hover {
  background-color: #dedede;
}
