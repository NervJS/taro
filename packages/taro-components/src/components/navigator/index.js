import Taro from '@tarojs/taro-h5'
import classNames from 'classnames'
import Nerv from 'nervjs'

import hoverable from '../../utils/hoverable'

import './navigator.css'

/* eslint-disable prefer-promise-reject-errors */

/**
 * Navigator组件参数
 * @typedef NavigatorProps
 * @property {String} appId 当target="miniProgram"时有效，要打开的小程序 appId
 * @property {String} ariaLabel 无障碍访问，（属性）元素的额外描述
 * @property {Number} delta 当 openType 为 'navigateBack' 时有效，表示回退的层数
 * @property {Object} extraData 当target="miniProgram"时有效，需要传递给目标小程序的数据，目标小程序可在 App.onLaunch()，App.onShow() 中获取到这份数据。详情
 * @property {String} [openType=navigate]  跳转方式
 * @property {String} path 当target="miniProgram"时有效，打开的页面路径，如果为空则打开首页
 * @property {String} [target=self]  在哪个目标上发生跳转，默认当前小程序，可选值self/miniProgram
 * @property {String} url 当前小程序内的跳转链接
 * @property {version} [version=release]  当target="miniProgram"时有效，要打开的小程序版本，有效值 develop（开发版），trial（体验版），release（正式版），仅在当前小程序为开发版或体验版时此参数有效；如果当前小程序是 *式版，则打开的小程序必定是正式版。
 * @property {String} onFail 当target="miniProgram"时有效，跳转小程序失败
 * @property {String} onComplete 当target="miniProgram"时有效，跳转小程序完成
 * @property {String} onSuccess 当target="miniProgram"时有效，跳转小程序成功
 */

/**
 * TODO: 参数还需要进一步细化对齐
 * Navigator组件
 * https://developers.weixin.qq.com/miniprogram/dev/component/navigator.html
 **/

@hoverable({
  hoverClass: 'navigator-hover',
  hoverStopPropergation: false,
  hoverStartTime: 50,
  hoverStayTime: 600
})
class Navigator extends Taro.Component {
  /** @type {NavigatorProps} */
  static defaultProps = {
    target: 'self',
    url: null,
    openType: 'navigate',
    delta: null,
    appId: null,
    path: null,
    extraData: {},
    version: 'release',
    onSuccess: null,
    onFail: null,
    onComplete: null,
    isHover: false
  }
  onClick = () => {
    const { openType, onSuccess, onFail, onComplete } = this.props
    let promise
    switch (openType) {
      case 'navigate':
        promise = Taro.navigateTo({
          url: this.props.url
        })
        break
      case 'redirect':
        promise = Taro.redirectTo({
          url: this.props.url
        })
        break
      case 'switchTab':
        promise = Taro.switchTab({
          url: this.props.url
        })
        break
      case 'reLaunch':
        promise = Taro.reLaunch({
          url: this.props.url
        })
        break
      case 'navigateBack':
        promise = Taro.navigateBack({
          delta: this.props.delta
        })
        break
      case 'exit':
        promise = Promise.reject({
          errMsg: `navigator:fail 暂不支持"openType: exit"`
        })
        break
    }
    if (promise) {
      promise.then(res => {
        onSuccess && onSuccess(res)
        onComplete && onComplete(res)
      }).catch(res => {
        onFail && onFail(res)
        onComplete && onComplete(res)
      })
    }
  }
  render () {
    const { isHover, hoverClass, onTouchStart, onTouchEnd, className } = this.props
    return (
      <div
        className={classNames(className, {
          [hoverClass]: isHover
        })}
        onTouchStart={onTouchStart}
        onTouchEnd={onTouchEnd}
        onClick={this.onClick}>
        {this.props.children}
      </div>
    )
  }
}

export default Navigator
