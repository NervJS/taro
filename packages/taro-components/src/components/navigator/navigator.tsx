import { Component, Prop, h, ComponentInterface, Host, Listen, Event, EventEmitter } from '@stencil/core'
import classNames from 'classnames'
const Taro = require('@tarojs/taro')

/**
 * Navigator组件参数
 * @typedef NavigatorProps
 * @property {String} appId 当target="miniProgram"时有效，要打开的小程序 appId
 * @property {String} ariaLabel 无障碍访问，（属性）元素的额外描述
 * @property {Number} delta 当 openType 为 'navigateBack' 时有效，表示回退的层数
 * @property {Object} extraData 当target="miniProgram"时有效，需要传递给目标小程序的数据，目标小程序可在 App.onLaunch()，App.onShow() 中获取到这份数据。详情
 * @property {String} [openType=navigate]  跳转方式
 * @property {String} path 当target="miniProgram"时有效，打开的页面路径，如果为空则打开首页
 * @property {String} [target=self]  在哪个目标上发生跳转，默认当前小程序，可选值self/miniProgram
 * @property {String} url 当前小程序内的跳转链接
 * @property {version} [version=release]  当target="miniProgram"时有效，要打开的小程序版本，有效值 develop（开发版），trial（体验版），release（正式版），仅在当前小程序为开发版或体验版时此参数有效；如果当前小程序是 *式版，则打开的小程序必定是正式版。
 * @property {String} onFail 当target="miniProgram"时有效，跳转小程序失败
 * @property {String} onComplete 当target="miniProgram"时有效，跳转小程序完成
 * @property {String} onSuccess 当target="miniProgram"时有效，跳转小程序成功
 */

/**
 * TODO: 参数还需要进一步细化对齐
 * Navigator组件
 * https://developers.weixin.qq.com/miniprogram/dev/component/navigator.html
 **/

@Component({
  tag: 'taro-navigator-core',
  styleUrl: 'navigator.css'
})
export class Navigator implements ComponentInterface {
  @Prop() hoverClass: string
  @Prop() url: string
  @Prop() openType = 'navigate'
  @Prop() isHover = false
  @Prop() delta = 0

  @Event({
    eventName: 'cuccess'
  }) onSuccess: EventEmitter

  @Event({
    eventName: 'fail'
  }) onFail: EventEmitter

  @Event({
    eventName: 'Complete'
  }) onComplete: EventEmitter

  @Listen('click')
  onClick () {
    const { openType, onSuccess, onFail, onComplete } = this
    let promise: Promise<any> = Promise.resolve()
    switch (openType) {
      case 'navigate':
        promise = Taro.navigateTo({
          url: this.url
        })
        break
      case 'redirect':
        promise = Taro.redirectTo({
          url: this.url
        })
        break
      case 'switchTab':
        promise = Taro.switchTab({
          url: this.url
        })
        break
      case 'reLaunch':
        promise = Taro.reLaunch({
          url: this.url
        })
        break
      case 'navigateBack':
        promise = Taro.navigateBack({
          delta: this.delta
        })
        break
      case 'exit':
        promise = Promise.reject(new Error('navigator:fail 暂不支持"openType: exit"'))
        break
    }
    if (promise) {
      promise.then(res => {
        onSuccess.emit(res)
      }).catch(res => {
        onFail.emit(res)
      }).finally(() => {
        onComplete.emit()
      })
    }
  }

  render () {
    const { isHover, hoverClass } = this

    return (
      <Host
        class={classNames({
          [hoverClass]: isHover
        })}
      />
    )
  }
}
