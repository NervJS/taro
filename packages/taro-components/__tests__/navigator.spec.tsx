import { h } from '@stencil/core'
import { newSpecPage, SpecPage } from '@stencil/core/testing'

import { Navigator } from '../src/components/navigator/navigator'

describe('Navigator', () => {
  let page: SpecPage

  it('base', async () => {
    page = await newSpecPage({
      components: [Navigator],
      template: () => (<taro-navigator-core />),
    })
    expect(page.root).toMatchSnapshot()
  })
})
