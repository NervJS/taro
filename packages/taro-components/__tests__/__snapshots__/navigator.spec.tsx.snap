// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigator base 1`] = `<taro-navigator-core></taro-navigator-core>`;
